server/src/main/java/org/elasticsearch/cluster/service/MasterService.java

```diff
@@ -296,7 +307,92 @@ private void runTasks(TaskInputs taskInputs) {
                 }
 
                 logger.debug("publishing cluster state version [{}]", newClusterState.version());
-                publish(clusterStatePublicationEvent, taskOutputs);
+                publish(
+                    clusterStatePublicationEvent,
+                    new CompositeTaskAckListener(
+                        executionResults.stream()
+                            .map(ExecutionResult::getContextPreservingAckListener)
+                            .filter(Objects::nonNull)
+                            .map(
+                                contextPreservingAckListener -> new TaskAckListener(
+                                    contextPreservingAckListener,
+                                    newClusterState.version(),
+                                    newClusterState.nodes(),
+                                    threadPool
+                                )
+                            )
+                            .collect(Collectors.toList())
```

Comment:
I believe this was moved from `createAckListener`, however is it possible to update it to use immutable list? 
```suggestion
                            .toList()
```

Reply:
Good point, yes.